README.md
# Factom.js

[![npm version](https://badge.fury.io/js/factom.svg)](https://badge.fury.io/js/factom)
[![Build Status](https://travis-ci.com/PaulBernier/factomjs.svg?branch=master)](https://travis-ci.com/PaulBernier/factomjs)
[![Coverage Status](https://coveralls.io/repos/github/PaulBernier/factomjs/badge.svg?branch=master)](https://coveralls.io/github/PaulBernier/factomjs?branch=master)
[![Known Vulnerabilities](https://snyk.io/test/github/PaulBernier/factomjs/badge.svg?targetFile=package.json)](https://snyk.io/test/github/PaulBernier/factomjs?targetFile=package.json)

Factom.js is a library to easily build applications on the Factom blockchain.

-   Provides higher level functionalities than factomd API making it a breeze to create chains or entries, send transactions...
-   Provides uniform, consistent and easy to use Factom data structures.
-   Provides a range of util functions to manipulate Factom addresses.

## Installation

```bash
$ npm install --save factom
```

## Documentation and changelog

The complete documentation of the library is available in Markdown format in the `docs` directory or online in web format at [https://factomjs.luciap.ca](https://factomjs.luciap.ca).

A changelog is available in the file `changelog.md`.

## Web browser

Two versions of _factom.js_ are being bundled for usage in a web browser and can be found in the `dist` folder. `dist/factom.js` is a bundle containing all the exposed components of the library. `dist/factom-struct.js` is a lighter bundle that contains factom structures such as Entry, Chain and Transaction, and all the utily functions related to FCT/EC addresses and fundamental constants. `factom-struct` bundle doesn't include any component that makes API calls.

```javascript
import { isValidPrivateAddress, isValidPrivateEcAddress } from 'factom/dist/factom-struct';
import { FactomCli } from 'factom/dist/factom';
```

## Usage

We recommend you have a look at the [tutorial on the developer portal of the Factom protocol](https://developers.factomprotocol.org/start/hello-world-examples/javascript).

**Important:** please note than whenever a private (Entry Credit or Factoid) address is needed in this library (typically for signing data), you can either provide a private address or a public address as an argument. If you provide a public address the library will attempt to retrieve the corresponding private address from the wallet. Thus providing private address as arguments allow you to not have to run walletd.

### Instantiate FactomCli

```javascript
const { FactomCli } = require('factom');

// Default factomd connection to localhost:8088
// and walletd connection to localhost:8089
const cli = new FactomCli();

// You can override factomd connection parameters and retry strategy
const cli = new FactomCli({
    host: '52.202.51.228',
    port: 8088,
    path: '/v2', // Path to V2 API. Default to /v2
    debugPath: '/debug', // Path to debug API. Default to /debug
    user: 'paul', // RPC basic authentication
    password: 'pwd',
    protocol: 'http', // http or https. Default to http
    rejectUnauthorized: true, // Set to false to allow connection to a node with a self-signed certificate
    timeout: 5000, // Timeout of individual requests
    retry: {
        retries: 3,
        factor: 2,
        minTimeout: 500,
        maxTimeout: 2000,
    },
});

// You can also override both factomd and walletd options
const cli = new FactomCli({
    factomd: {
        host: '52.202.51.228',
        port: 8088,
    },
    walletd: {
        host: '52.202.51.228',
        port: 8089,
    },
});
```

### Chains and Entries

#### Add a Chain

```javascript
const { Entry, Chain } = require('factom');
const firstEntry = Entry.builder()
    .extId('6d79206578742069642031') // If no encoding parameter is passed as 2nd argument, 'hex' is used
    .extId('my ext id 1', 'utf8') // Explicit the encoding. Or you can pass directly a Buffer
    .content('Initial content', 'utf8')
    .build();

const chain = new Chain(firstEntry);
cli.add(chain, 'Es32PjobTxPTd73dohEFRegMFRLv3X5WZ4FXEwNN8kE2pMDfeMym');
```

#### Add an entry

```javascript
const { Entry } = require('factom');
const myEntry = Entry.builder()
    .chainId('9107a308f91fd7962fecb321fdadeb37e2ca7d456f1d99d24280136c0afd55f2')
    .extId('6d79206578742069642031') // If no encoding parameter is passed as 2nd argument, 'hex' is used
    .extId('some external ID', 'utf8')
    .content('My new content', 'utf8')
    .build();
cli.add(myEntry, 'Es32PjobTxPTd73dohEFRegMFRLv3X5WZ4FXEwNN8kE2pMDfeMym');
```

```javascript
// Add multiples entries.
// The library will limit the number of concurrent Promises to 200 by default to avoid overwhelming the factomd instance.
cli.add([entry1, entry2], 'Es32PjobTxPTd73dohEFRegMFRLv3X5WZ4FXEwNN8kE2pMDfeMym');

// The concurrency limit can be customized.
cli.add([entry1, entry2], 'Es32PjobTxPTd73dohEFRegMFRLv3X5WZ4FXEwNN8kE2pMDfeMym', {
    concurrency: 1,
});
```

#### Commit/reveal acknowledgment when submitting chains or entries

Factom protocol uses a [commit/reveal commitment scheme](https://en.wikipedia.org/wiki/Commitment_scheme). By default when using `add` the library will sequentially wait for an acknowledgment (ack) of the commit by the network and then wait for the ack of the reveal, both for up to 60s. The library allows you to customize the timeouts of those acks and also to not wait for the acks at all when creating a Chain or an Entry. Please read [Factom whitepaper](https://www.factom.com/devs/docs/guide/factom-white-paper-1-0) about commit/reveal scheme and what are the potential risks to not wait for network acknowledgments.

```javascript
// Default behavior waits for both commit and reveal up to 60s
cli.add(myEntry, 'Es32PjobTxPTd73dohEFRegMFRLv3X5WZ4FXEwNN8kE2pMDfeMym');
// Change the timeout for commit ack to 120s and the timeout for reveal ack to 20s
cli.add(myEntry, 'Es32PjobTxPTd73dohEFRegMFRLv3X5WZ4FXEwNN8kE2pMDfeMym', {
    commitTimeout: 120,
    revealTimeout: 20,
});
// By providing a negative number the library will not wait for any acknowledgment.
// In below example the wait on reveal ack is disabled (it'll still wait up to 60s on the commit ack).
cli.add(myEntry, 'Es32PjobTxPTd73dohEFRegMFRLv3X5WZ4FXEwNN8kE2pMDfeMym', { revealTimeout: -1 });
```

#### Repeated commit

If you commit twice an entry or a chain and that the second time the fees paid are lower or equal to the first commit you are in a 'repeated commit' case and the second commit will be rejected (and you won't be charged for it). If this scenario happens the output of `add` will have the field `repeatedCommit` set to `true` and the field `txId` will be undefined. See [Factom doc](https://docs.factom.com/api#repeated-commit).

#### Getting entries and block context

The simplest and fastest way to retrieve an Entry is to query it by its hash.

```javascript
cli.getEntry('caf017da212bb68ffee2ba645e1488e5834863743d50972dd3009eab2b93eb42');
```

You may notice that the entry returned doesn't have a timestamp populated. That's because the timestamp is actually stored in the Entry Block that contains this Entry. The library offers a convenient way to also retrieve that information:

```javascript
cli.getEntryWithBlockContext('caf017da212bb68ffee2ba645e1488e5834863743d50972dd3009eab2b93eb42');
```

The entry returned will have its timestamp populated and an additional `blockContext` field that is structured as follow:

```javascript
// Note that the timestamps here are in epoch seconds whereas the timestamp attribute of the Entry itself is in timestamp milliseconds
{ entryTimestamp: 1518286500,
  directoryBlockHeight: 7042,
  entryBlockTimestamp: 1518286440,
  entryBlockSequenceNumber: 1,
  entryBlockKeyMR: 'a13ac9df4153903f5a07093effe6434bdeb35fea0ff4bd402f323e486bea6ea4' }
```

Besides `getEntryWithBlockContext` entries returned by `getFirstEntry` and `getAllEntriesOfChain` also have a `timestamp` and `blockContext` populated.

#### Iterating entries of a chain

The FactomCli method `getAllEntriesOfChain` fetches all the entries of the chain before returning the result: in case of long chains in can be impractical. In some cases you may want to iterate only through a portion of the chain. For those cases FactomCli exposes the method `rewindChainWhile(chainId, function predicate(entry) {}, function body(entry) {})` that iterates a chain from the most recent entry to the oldest one as long as the `predicate` function returns true and that the end of the chain has not been reached. At each iteration the `body` function is called with the current entry as its argument.

_Example 1. Iterating a long chain entry by entry_

```javascript
await cli.rewindChainWhile('caf017da212bb68ffee2ba645e1488e5834863743d50972dd3009eab2b93eb42',
    () => true,
    entry => // Process entry
```

_Example 2. Searching an entry in a chain_

```javascript
let search = true,
    found;
await cli.rewindChainWhile(
    'caf017da212bb68ffee2ba645e1488e5834863743d50972dd3009eab2b93eb42',
    () => search,
    function (entry) {
        if (entry.extId[0].toString() === 'Find me!') {
            search = false;
            found = entry;
        }
    }
);
```

### Addresses

Factom.js offers a bunch of util functions around FCT/EC addresses and cryptographic keys, namely:

```javascript
const {
    isValidAddress,
    addressToKey, // For EC, Es, Fs addresse
    addressToRcdHash, // For FA addresses
    isValidPublicAddress,
    isValidPrivateAddress,
    isValidEcAddress,
    isValidPublicEcAddress,
    isValidPrivateEcAddress,
    isValidFctAddress,
    isValidPublicFctAddress,
    isValidPrivateFctAddress,
    getPublicAddress,
    keyToPublicFctAddress,
    rcdHashToPublicFctAddress,
    seedToPrivateFctAddress,
    keyToPublicEcAddress,
    seedToPrivateEcAddress,
    generateRandomFctAddress,
    generateRandomEcAddress,
} = require('factom');
```

### Transactions

#### Transaction object

Note that all the amounts are in factoshis (10^-8 Factoids).

```javascript
Transaction {
    timestamp: 1527092064498,
    inputs: [TransactionAddress {
        address: 'FA3syRxpYEvFFvoN4ZfNRJVQdumLpTK4CMmMUFmKGeqyTNgsg4uH',// Paying FCT address
        amount: 12000, // Amount in factoshis
        rcdHash: < Buffer fb cc 9 b c3 02 cc b8 c3 0 c 69 0 c 70 e4 12 f0 05 53 cc f8 5e 4 b 6 c 2 a b6 0e ce db 12 fe 9 d fe aa >
    }],
    factoidOutputs: [TransactionAddress {
        address: 'FA3cnxxcRxm6RQs2hpExdEPo9utyeBZecWKeKa1pFDCrRoQh9aVw',// Receiving FCT address
        amount: 1, // Amount in factoshis
        rcdHash: < Buffer d9 54 88 34 81 f3 aa 50 1 f 3 f 5 d 4 f 9e 79 6 b af e8 aa 01 bf e8 97 80 77 1e 73 3 d 63 96 f8 fb 9 b >
    }],
    entryCreditOutputs: [TransactionAddress {
        address: 'EC3MVTBYTo2Y1HrEKxeEGfNNoKhLZ9ZYQhb26zQUzngJ6SLUVRX9',// Receiving EC address
        amount: 10000, // Amount in factoshis
        rcdHash: < Buffer d1 aa eb 70 6 c 79 47 5 d b8 8e 01 d9 e4 17 e7 83 2 b 20 df 5 c c6 fd a6 7 f 09 b8 8 b 89 36 64 2 a 9 f >
    }],
    marshalBinarySig: < Buffer 02 01 63 8 d c7 b0 f2 01 01 01 dd 60 fb cc 9 b c3 02 cc b8 c3 0 c 69 0 c 70 e4 12 f0 05 53 cc f8 5e 4 b 6 c 2 a b6 0e ce db 12 fe 9 d fe aa 01 d9 54 88 34 81... > ,
    id: '40dee7fde9747e4b4a8e9d4685c64044d1a7513734f6adf63698b1533b57461d',

    rcds: [ < Buffer 01 1 b cb 4 c 8 a 77 1 c 28 69 dd f5 54 65 54 14 e5 6 b df 36 06 63 f3 39 60 03 9 a 9 a a4 3 a c5 82 03 06 > ],
    signatures: [ < Buffer 4 b d5 bc 2 a 4 d e0 a0 06 2 b 30 28 d3 34 90 31 f7 e4 93 5e a2 6 a db 20 f1 0e c8 92 9 c db 7e 62 f2 c4 9 b 0 f 14 e5 cd 6 d fe 28 22 1e c1 9 a bb 32 aa 70 83... > ],

    totalInputs: 12000, // Sum of all the inputs in factoshis
    totalFactoidOutputs: 1, // Sum of all factoid outputs in factoshis
    totalEntryCreditOutputs: 10000, // Sum of all entry credit outputs in factoshis
    feesPaid: 11999 // Fees paid by this transaction in factoshis
}

```

#### Simple Factoid transaction

```javascript
// Send 1000000 Factoshis (10^-8 Factoids) from Fs2w6VL6cwBqt6SpUyPLvdo9TK834gCr52Y225z8C5aHPAFav36X
// to FA3cnxxcRxm6RQs2hpExdEPo9utyeBZecWKeKa1pFDCrRoQh9aVw
const transaction = await cli.createFactoidTransaction(
    'Fs2w6VL6cwBqt6SpUyPLvdo9TK834gCr52Y225z8C5aHPAFav36X',
    'FA3cnxxcRxm6RQs2hpExdEPo9utyeBZecWKeKa1pFDCrRoQh9aVw',
    1000000
);
// You can check the additional fees that you are going to pay for the transaction to go through
// (automatically set to the minimum acceptable by the network)
console.log(transaction.feesPaid);
// Send the transaction
const txId = await cli.sendTransaction(transaction);
```

#### Buy EntryCredit

```javascript
// Buy 10 EC with address Fs2w6VL6cwBqt6SpUyPLvdo9TK834gCr52Y225z8C5aHPAFav36X and credited
// to EC2UFobcsWom2NvyNDN67Q8eTdpCQvwYe327ZeGTLXbYaZ56e9QR
const transaction = await cli.createEntryCreditPurchaseTransaction(
    'Fs2w6VL6cwBqt6SpUyPLvdo9TK834gCr52Y225z8C5aHPAFav36X',
    'EC2UFobcsWom2NvyNDN67Q8eTdpCQvwYe327ZeGTLXbYaZ56e9QR',
    10
);
// You can check how many Factoshis it's going to cost you to buy those 10 EC
console.log(transaction.totalInputs);
const txId = await cli.sendTransaction(transaction);
```

#### Multi inputs/outputs transaction

For multi inputs/outputs you have to build your Transaction object yourself and set the fees manually.

```javascript
const { Transaction } = require('factom');

const ecRate = await cli.getEntryCreditRate();
const tmpTx = Transaction.builder()
    .input('Fs2w6VL6cwBqt6SpUyPLvdo9TK834gCr52Y225z8C5aHPAFav36X', 14000000)

changelog.md
# Changelog

## [1.4.1]

Upgrade dependencies.

## [1.4.0]

### Added

-   Add support for pending Factoid transactions to `FactomEventEmitter`.

## [1.3.0]

### Fixed

-   Fix an incorrect behavior of `FactomEventEmitter` that would immediately emit an event when the first listener was attached. This can be a breaking change if you are relying on this initial event to be emitted.
-   Fix TypeScript declaration file.

## [1.2.7]

### Added

-   Pass the paying EC public address as the second argument to the signing function when using signature delegation.

## [1.2.6]

### Added

-   `ApiError` now has a `data` field populated with the JSON-RPC error `data` field.

## [1.2.4]

### Added

-   Basic support of cookies. This allows session affinity for Factom Open Node (currently uses a Cloudflare load balancer).

## [1.2.3]

### Added

-   Added `retry` field to the `requestConfig` parameter of `FactomCli.factomdApi`, `FactomCli.walletdApi`, `FactomdCli.call` and `WalletdCli.call`. Allows to override the instance level retry strategy for a single API call.

## [1.2.0]

### Added

-   Added `timeout` option to `ConnectionOptions`. Allows to set a timeout for all API requests made by `FactomCli`, `FactomdCli` and `WalletdCli`.
-   Added a 3rd optional `requestConfig` parameter to `FactomCli.factomdApi`, `FactomCli.walletdApi`, `FactomdCli.call` and `WalletdCli.call`. Currenly only allows to specify a request timeout.

## [1.1.1]

### Fixed

-   Fix `FactomEventEmitter` typescript definition.

## [1.1.0]

### Added

-   New `FactomEventEmitter` that can emit various blockchain events (see doc for details).
-   New methods `composeChainCommitDelegateSig`, `composeChainDelegateSig`, `composeEntryCommitDelegateSig` and `composeEntryDelegateSig` for delegating signing of chain and entry commits to an external function (usage for hardware wallet).
-   `FactomCli#add*` and `FactomCli#commit*` methods now support delegating signing of the commit messages by passing a `sign` function in the options.

### Deprecated

-   `chunkSize` option for the `FactomCli#add*` methods has been deprecated and replaced by `concurrency`.
-   Depracated `signature` third argument of functions `composeChainCommit`, `composeChain`, `composeEntryCommit` and `composeEntryCommit`. Use the new `*DelegateSig` methods instead.

## [1.0.2]

### Removed

-   Revert the unnecessary chanage in 1.0.1 for `rewindChainWhile` (the entry already contains a blockContext).

## [1.0.1]

### Added

-   BlockContext populated for transactions in a `FactoidBlock` object (except `directoryBlockKeyMR` that is not available).
-   `rewindChainWhile(chainId, predicate, fn)`: the `predicate` and `fn` functions now receive the entry block header as a second argument.

## [1.0.0]

### Added

-   JSDoc of all public classes and methods.
-   New util functions `generateRandomFctAddress` and `generateRandomEcAddress`.
-   Method `toObject()` added to `Entry` and `Chain` classes. Returns a JavaScript object than can be serialized.
-   TypeScript definition file (RewardChain contribution).

### Fixed

-   Catch repeated commit exceptions properly.
-   Rejection of negative transaction fees.
-   Handling of CORS with credentials.
-   Handle case `Error.captureStackTrace` is undefined (Firefox).
-   Fix bug preventing to instanciate `FactomdCli` without params.

### Changed

-   More explicit ack errors.
-   `getTransaction` throws an `Error` if no transaction is found instead of returning `undefined`. To be consistent with `getEntry` behavior.
-   Redirect factomd `current-minute` API call to v2 path (intead of debug path).
-   Increased performance of `getEntryWithBlockContext`: fixed cost of 3 factomd API calls instead of rewinding the entry blocks of the chain.
-   `keyToPrivateFctAddress` renamed `seedToPrivateFctAddress`.
-   `keyToPrivateEcAddress` renamed `seedToPrivateEcAddress`.
-   `add` throws an Error when trying to create a chain that already exists.
-   All the `add*` functions validate by default that the paying EC address holds enough funds to pay for the commits. Can be bypassed by setting the option `skipFundValidation` to `true`.
-   Added a check before sending a transaction that all EC ouputs have a factoshi amount above the minimum amount required to get at least 1 EC. It can be bypassed using the `force` option of `sendTransaction` method.
-   Renamed `isValidFctPublicAddress`, `isValidFctPrivateAddress`, `isValidEcPrivateAddress` and `isValidEcPublicAddress` to respectively `isValidPublicFctAddress`, `isValidPrivateFctAddress`, `isValidPrivateEcAddress` and `isValidPublicEcAddress`.

### Removed

-   Remove `FactomCli` methods `getNodeProperties` and `getWalletProperties`. Unnecessary wrappers.
-   Do not expose internal constants anymore.

## [0.3.6] 2018-10-03

### Fixed

-   Use Babel plugin `@babel/transform-modules-commonjs` to fix integration in the browser (broke after upgrade to Babel 7).

## [0.3.5] 2018-09-26

### Fixed

-   Parsing of the descriptor index in an Admin block coinbase cancel entry.

## [0.3.4] 2018-09-13

### Added

-   New FacomCli method `rewindChainWhile` that allows to iterate on a chain entry by entry.
-   Expose `composeChainLedger` and `composeEntryLedger` functions.
-   Allow manual signature of entry and chain commits (hardware wallet integration): `composeChainCommit` and `composeEntryCommit` can now take an EC public address as the 2nd argument and the commit signature as the 3rd argument.

## [0.3.3] 2018-09-10

### Added

-   Options to override default API paths (`/v2` and `/debug`).

## [0.3.2] 2018-09-05

### Changed

-   Renamed `addressToRcd` to `addressToRcdHash`. (breaking change)
-   Fixed README.

## [0.3.1] 2018-08-29

### Changed

-   Upgrade to Babel 7 for the bundling process.

## [0.3] 2018-08-27

### Added


factom.js
const factomdjs = require('../factomdjs/src/factomd'),
    walletd = require('../factom-walletdjs/src/factom-walletd'),
    log = require('winston'),
    EdDSA = require('elliptic').eddsa,
    fctUtils = require('factomjs-util'),
    Promise = require('bluebird'),
    crypto = require('crypto');

const ec = new EdDSA('ed25519');

const CHAIN_CREATION_COST = 10;
const NULL_HASH = '0000000000000000000000000000000000000000000000000000000000000000';

function setFactomNode(url) {
    factomdjs.setFactomNode(url);
}

function setFactomWallet(url) {
    walletd.setFactomNode(url);
}

// MIGRATED
// TODO: Paginated version
async function getAllEntriesOfChain(chainId) {
    const allEntries = [];
    const chainHead = await factomdjs.chainHead(chainId);

    let keymr = chainHead.chainhead;
    while (keymr !== NULL_HASH) {
        const {
            entries,
            prevkeymr
        } = await getAllEntriesOfEntryBlock(keymr);
        allEntries.push(...entries.reverse());

        keymr = prevkeymr;
    }

    return Promise.resolve(allEntries.reverse());
}

// MIGRATED
async function getAllEntriesOfEntryBlock(keymr) {
    const entryBlock = await factomdjs.entryBlock(keymr);

    const entries = await Promise.map(entryBlock.entrylist.map(e => e.entryhash), factomdjs.entry);

    return {
        entries: entries.map(decodeEntry),
        prevkeymr: entryBlock.header.prevkeymr
    };
}

// MIGRATED
async function getFirstEntry(chainId) {
    const chainHead = await factomdjs.chainHead(chainId);
    let keymr = chainHead.chainhead;
    let entryBlock;
    while (keymr !== NULL_HASH) {
        entryBlock = await factomdjs.entryBlock(keymr);
        keymr = entryBlock.header.prevkeymr;
    }

    return factomdjs.entry(entryBlock.entrylist[0].entryhash)
        .then(decodeEntry);
}

// Migrated
function getChainHead(chainId) {
    return factomdjs.chainHead(chainId);
}

function decodeEntry(entry) {
    const content = entry.content ? Buffer.from(entry.content, 'hex') : Buffer.from('');
    const extIds = Array.isArray(entry.extids) ? entry.extids.map(extid => Buffer.from(extid, 'hex')) : [];

    return {
        chainId: entry.chainid,
        content: content,
        extIds: extIds
    };
}

// MIGRATED
async function chainExists(chainId) {
    return factomdjs.chainHead(chainId)
        .then(() => true)
        .catch(() => false);
}

// MIGRATED
function getChainId(firstEntry) {
    const extIdsHashes = toBuffers(firstEntry.extIds).map(sha256);
    const hashes = Buffer.concat(extIdsHashes);

    return sha256(hashes);
}

// MIGRATED
// TODO: addEntry and addChain are exactly the same besides composeChain/commitChain
// Handle both cases argument is chain or just firstEntry
// TODO: safe/unsage version
async function addChain(firstEntry, ecAddress) {
    if (!fctUtils.isValidAddress(ecAddress)) {
        throw `${ecAddress} is not a valid address`;
    }

    let composed = {};
    if (ecAddress.substring(0, 2) === 'EC') {
        const {commit, reveal} = await walletd.composeChain(firstEntry.extIds.map(toHex), toHex(firstEntry.content), ecAddress);
        composed.commit = commit.params.message;
        composed.reveal = reveal.params.entry;
    } else if(ecAddress.substring(0, 2) === 'Es') {
        const {commit, reveal} = composeChain(firstEntry, ecAddress);
        composed.commit = commit.toString('hex');
        composed.reveal = reveal.toString('hex');
    } else {
        throw `${ecAddress} is not an EC address`;
    }

    const commitPromise = factomdjs.commitChain(composed.commit).catch(function (e) {
        if (e.message === 'Repeated Commit') {
            log.warn(e);
        } else {
            throw e;
        }
    });

    const [committed, revealed] = await Promise.all([
        commitPromise,
        factomdjs.revealChain(composed.reveal)
    ]);

    log.debug(committed);
    log.debug(revealed);

    return {
        chainId: revealed.chainid,
        entryHash: revealed.entryhash
    };
}

// MIGRATED
// TODO: safe/unsage version
async function addEntry(entry, ecAddress) {
    if (!fctUtils.isValidAddress(ecAddress)) {
        throw `${ecAddress} is not a valid address`;
    }

    let composed = {};
    if (ecAddress.substring(0, 2) === 'EC') {
        const {commit, reveal} = await walletd.composeEntry(entry.chainId, entry.extIds.map(toHex), toHex(entry.content), ecAddress);
        composed.commit = commit.params.message;
        composed.reveal = reveal.params.entry;
    } else if(ecAddress.substring(0, 2) === 'Es') {
        const {commit, reveal} = composeEntry(entry, ecAddress);
        composed.commit = commit.toString('hex');
        composed.reveal = reveal.toString('hex');
    } else {
        throw `${ecAddress} is not an EC address`;
    }

    const commitPromise = factomdjs.commitEntry(composed.commit).catch(function (e) {
        if (e.message === 'Repeated Commit') {
            log.warn(e);
        } else {
            throw e;
        }
    });

    const [committed, revealed] = await Promise.all([
        commitPromise,
        factomdjs.revealEntry(composed.reveal)
    ]);

    log.debug(committed);
    log.debug(revealed);

    return Promise.resolve(revealed.entryhash);
}

// MIGRATED
async function addEntries(entries, ecpub) {
    return Promise.map(entries, entry => addEntry(entry, ecpub));
}

function toHex(input) {
    if (!input) {
        return '';
    }
    return Buffer.from(input).toString('hex');
}

// MIGRATED
function entrySize(entry) {
    // 35 = header size
    let size = 35;

    if (Array.isArray(entry.extIds)) {
        const extIdsBuffers = toBuffers(entry.extIds);
        const extIdsLength = Buffer.concat(extIdsBuffers).length;
        size += 2 * entry.extIds.length + extIdsLength;
    }

    if (entry.content) {
        size += entry.content.length;
    }

    return size;
}

// MIGRATED
function chainCost(firstEntry) {
    return 10 + entryCost(firstEntry);
}

// MIGRATED
function entryCost(entry) {
    // Header size (35) is not counted in the cost
    const dataLength = entrySize(entry) - 35;
    if (dataLength > 10240) {
        throw 'Entry cannot be larger than 10Kb';
    }

    return Math.ceil(dataLength / 1024);
}

// MIGRATED
function waitOnCommitAck(txid, timeout) {
    return waitOnAck(txid, 'c', 'commitdata', timeout);
}

// MIGRATED
function waitOnRevealAck(hash, chainId, timeout) {
    return waitOnAck(hash, chainId, 'entrydata', timeout);
}

// MIGRATED
function waitOnAck(hash, chainId, ackResponseField, to) {
    if (!hash || !chainId) {
        return Promise.reject('Invalid argument: hash or chain ID is missing');
    }

    const timeout = to || 60;
    const startTime = Date.now();

    return new Promise((resolve, reject) => {
        const clearId = setInterval(async function () {
            process.stdout.write('.');
            const ackResponse = await factomdjs.ack(hash, chainId).catch(function (e) {
                clearInterval(clearId);
                process.stdout.write('\n');
                reject(e);
            });
            const status = ackResponse[ackResponseField].status;

            if (status !== 'Unknown' && status !== 'NotConfirmed') {
                clearInterval(clearId);
                process.stdout.write('\n');
                resolve(status);
            }

            if ((Date.now() - startTime) > timeout * 1000) {
                clearInterval(clearId);
                process.stdout.write('\n');
                reject('Ack timeout');
            }

        }, 500);
    });
}

// MIGRATED
function properties() {
    return factomdjs.properties();
}

// MIGRATED
function getBalance(x) {
    // TODO: detect type of X and redirect
    // TODO: implement https://github.com/FactomProject/factom/blob/a0a55096f9d2aeb5cb63b8b5a714a285f3a100b3/addresses.go#L43
    return factomdjs.entryCreditBalance(x)
        .then(res => res.balance);
}

// MIGRATED
function composeChainCommit(firstEntry, ecPrivate) {
    const chainId = getChainId(firstEntry);
    const entryHash = getEntryHash(firstEntry);
    const cost = entryCost(firstEntry);
    const buffer = Buffer.alloc(104);

    buffer.writeInt8(0);
    buffer.writeIntBE(Date.now(), 1, 6);
    const chainIdHash = sha256d(chainId);
    chainIdHash.copy(buffer, 7);
    const commitWeld = sha256d(Buffer.concat([entryHash, chainId]));
    commitWeld.copy(buffer, 39);
    entryHash.copy(buffer, 71);
    buffer.writeInt8(cost + CHAIN_CREATION_COST, 103);

    // Signing commit
    const secret = fctUtils.privateHumanAddressStringToPrivate(ecPrivate);
    const key = ec.keyFromSecret(secret);
    const signature = Buffer.from(key.sign(buffer).toBytes());
    const ecPublic = Buffer.from(key.getPublic());

    return Buffer.concat([buffer, ecPublic, signature]);
}

// MIGRATED
function composeChainReveal(firstEntry) {
    return marshalEntryBinary(firstEntry);
}

// MIGRATED
function composeChain(firstEntry, ecPrivate) {
    firstEntry.chainId = getChainId(firstEntry);
    return {
        commit: composeChainCommit(firstEntry, ecPrivate),
        reveal: composeChainReveal(firstEntry)
    };
}

// MIGRATED
function composeEntryCommit(entry, ecPrivate) {
    const entryHash = getEntryHash(entry);
    const cost = entryCost(entry);
    const buffer = Buffer.alloc(40);

    buffer.writeInt8(0);
    buffer.writeIntBE(Date.now(), 1, 6);
    entryHash.copy(buffer, 7);
    buffer.writeInt8(cost, 39);

    // Signing commit
    const secret = fctUtils.privateHumanAddressStringToPrivate(ecPrivate);
    const key = ec.keyFromSecret(secret);
    const signature = Buffer.from(key.sign(buffer).toBytes());
    const ecPublic = Buffer.from(key.getPublic());

    return Buffer.concat([buffer, ecPublic, signature]);
}

// MIGRATED
function composeEntryReveal(entry) {
    return marshalEntryBinary(entry);
}

// MIGRATED
function composeEntry(entry, ecPrivate) {
    return {
        commit: composeEntryCommit(entry, ecPrivate),
        reveal: composeEntryReveal(entry)
    };
}

// MIGRATED
function getEntryHash(entry) {
    const data = marshalEntryBinary(entry);
    return sha256(Buffer.concat([sha512(data), data]));
}

function toBuffers(extIds) {
    if (!extIds) {
        return [];
    }
    return extIds.map(Buffer.from);
}
// MIGRATED
function marshalEntryBinary(entry) {
    const externalIds = marshalExternalIdsBinary(entry.extIds);
    const header = marshalHeaderBinary(entry.chainId, externalIds);
    const content = Buffer.from(entry.content ? entry.content : '');

    return Buffer.concat([header, externalIds, content]);
}
// MIGRATED
function marshalHeaderBinary(chainId, marshalledExtIds) {
    const chainIdBuffer = Buffer.isBuffer(chainId) ? chainId : Buffer.from(chainId, 'hex');
    const header = Buffer.alloc(35);
    chainIdBuffer.copy(header, 1);
    header.writeInt16BE(marshalledExtIds.length, 33);

    return header;
}
// MIGRATED
function marshalExternalIdsBinary(extIds) {
    const buffers = toBuffers(extIds);
    const result = [];

    for (let extId of buffers) {
        const size = Buffer.alloc(2);
        size.writeInt16BE(extId.length);
        result.push(size);
        result.push(extId);
    }

    return Buffer.concat(result);
}
// MIGRATED
function sha256(data) {
    const hash = crypto.createHash('sha256');
    hash.update(data);
    return hash.digest();
}
// MIGRATED
function sha256d(data) {
    return sha256(sha256(data));
}
// MIGRATED
function sha512(data) {
    const hash = crypto.createHash('sha512');
    hash.update(data);
    return hash.digest();
}

//setFactomNode('http://localhost:8088/v2')

// addChain({
//             extIds: ["hello-world8"],
//             content: "content"
//         },
//         "EC2vXWYkAPduo3oo2tPuzA44Tm7W6Cj7SeBr3fBnzswbG5rrkSTD"
//     )
//     .then(console.log)
//     .catch(console.error);

// const l = 1024 - 2 * 2 - 3 - 3;

// const str = new Array(l + 1).join('r');
// console.log(str.length)

// const entry = {
//     chainId: "ac44e52c539065efa3563104248c984ef893f0e3351d7c59c93bf022d2214e63",
//     extIds: ["arf", "arf"],
//     content: str
// };
// console.log(entryCost(entry))
// addEntry(entry, "EC2vXWYkAPduo3oo2tPuzA44Tm7W6Cj7SeBr3fBnzswbG5rrkSTD")
//     .then(entryHash => waitOnRevealAck(entryHash, entry.chainId))
//     .then(console.log)
//     .catch(console.error)

// addEntries([{
//         chainId: "332d06db89e9f4fbf7c7940c21ca7aa01a38f8c332d79dcf234538afa9d6b31a",
//         extIds: ["hello-world4"],
//         content: "my content"
//     }, {
//         chainId: "332d06db89e9f4fbf7c7940c21ca7aa01a38f8c332d79dcf234538afa9d6b31a",
//         extIds: ["hello-world6"],
//         content: "my content"
//     }], "EC2vXWYkAPduo3oo2tPuzA44Tm7W6Cj7SeBr3fBnzswbG5rrkSTD")
//     .then(console.log)
//     .catch(console.error);

// getFirstEntry('43649840f6342be91be137e46f447cffaa6796b7fe6f4fdd8acb80744ab1cd6c').then(console.log)

//waitOnRevealAck('831cc02713ae454045607bbb928644b3fd48ec342a2cc6c93dab8aad206bea8e', 'fbf1bb7ffa4ec0bbb0f7dc18cbeb47514102c2eb38fd1f985be3254156b28677');

module.exports = {
    setFactomNode,
    setFactomWallet,
    getAllEntriesOfChain,
    getAllEntriesOfEntryBlock,
    getFirstEntry,
    getChainHead,
    getChainId,
    properties,
    getBalance,
    chainExists,
    entrySize,
    entryCost,
    chainCost,
    addChain,
    addEntry,
    addEntries,
    waitOnRevealAck,
    waitOnCommitAck
};